# Laboratorio 4: Spark y EDA 🐼

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Catherine Benavides
- Ayudante: Nicolás Ojeda, Eduardo Moya

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Bárbara Aguayo
- Nombre de alumno 2: Camila Figueroa

### **Link de repositorio de GitHub:** `https://github.com/bluemili/MDS7202-Labs`

### Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 7 días desde la publicación, 3 días de atraso con 1 punto de descuento c/u.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos Y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquer material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pyspark` para manejar datos tabulares de gran volúmen.
- Crear gráficos para el desarrollo de Análisis de Datos Exploratorios (EDA).

### Datos del Lab

- Base de datos: https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/datos_lab_spark.parquet
- Objeto serializado: https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/object.pkl?inline=false

## Preguntas Teóricas [12 puntos]
(2 por pregunta)

<center>
<img src="https://img.buzzfeed.com/buzzfeed-static/static/2018-08/1/17/enhanced/buzzfeed-prod-web-05/anigif_enhanced-9173-1533160033-1.gif" width=350 />



Responda en  máximo 5 líneas las siguientes preguntas:
1. ¿Qué es Apache Spark y cuáles son sus principales ventajas sobre Pandas?
2. ¿Qué es un RDD en Spark? Describe una de sus principales características. ¿Qué tienen que ver con los dataframes?.
3. Diferencia entre transformaciones y acciones en Spark. Proporciona ejemplos de cada una. ¿Qué ocurre internamente cuando se ejecuta una acción?
4. Explica la importancia del particionamiento en Spark y cómo afecta el rendimiento del procesamiento de datos.
5. ¿Cuáles son las funciones de Spark Driver y Spark Executor?
6. ¿Qué es el Catalyst Optimizer en Apache Spark y cuál es su función principal en la optimización de consultas SQL?


**Respuestas**

**¿Qué es Apache Spark y cuáles son sus principales ventajas sobre Pandas?**

Apache Spark es un motor unificado para procesar datos a gran escala. Puede funcionar en centros físicos y en la nube.

Sus principales ventajas sobre Pandas son:
- Se desempeña de muchísima mejor forma al trabajar con volúmenes de datos grandes.
- Puede realizar operaciones distribuidas, siendo más eficiente y rápido que Pandas, sobre todo utilizando gran cantidad de datos.
- Tiene una gestión de memoria mucho más eficiente.

**¿Qué es un RDD en Spark? Describe una de sus principales características.**

Un Resilient Distributed Dataset (RDD) constituye el pilar fundamental de Apache Spark, siendo una estructura de datos distribuida e inmutable. Una de sus características principales es que los datos se encuentran divididos (o distribuidos) en particiones o nodos en el clúster.

**¿Qué tienen que ver con los dataframes?**

Los dataframes de PySpark corresponden a abstracciones de nivel superior de un RDD, es decir, son estructuras de datos más simples de utilizar para los usuarios, dado que ofrecen una interfaz similar a las tablas de bases de datos con filas y columnas.

**Diferencia entre transformaciones y acciones en Spark. Proporciona ejemplos de cada una. ¿Qué ocurre internamente cuando se ejecuta una acción?**

Las transformaciones generan un nuevo dataframe o conjunto de datos distribuidos, pero sin modificar el dataframe original. Esto con el fin de realizar tareas como filtrar, mapear y agrupar los datos.

En cambio, las acciones son activadoras que desencadenan la ejecución de las transformaciones necesarias para producir el resultado deseado.

**Explica la importancia del particionamiento en Spark y cómo afecta el rendimiento del procesamiento de datos.**

El particionamiento en Spark permite un rendimiento eficiente en el procesamiento de datos, debido a que cada partición se procesa de forma independiente y distribuidas en el clúster. De esta forma, es posible realizar tareas y operaciones de forma paralela, lo que permite equilibrar la carga de trabajo entre los nodos del cluster, optimizando el rendimiento y los tiempos de ejecución.

**¿Cuáles son las funciones de Spark Driver y Spark Executor?**

Las distintas particiones son coordinadas a partir del Spark Driver, el cual tiene la función de instanciar la SparkSession y comunicarse con el gestor del cluster para pedir recursos como memoria, CPU, entre otros.

Por otro lado, Spark Executor se encuentra en cada uno de los nodos del cluster y su función es realizar las tareas asignadas por el conductor en el nodo respectivo y enviar los resultados al mismo.

**¿Qué es el Catalyst Optimizer en Apache Spark y cuál es su función principal en la optimización de consultas SQL?**

El Catalyst Optimizer es una componente o herramienta fundamental de Spark SQL, que corresponde a una sección de Spark creada para ejecutar queries SQL. En esencia, corresponde a una biblioteca de árboles con instrucciones y cómo manipularlo.

La función principal de Catalyst Optimizer es mejorar el rendimiento de las consultas SQL al transformarlas en el plan más eficiente para ejecutar las operaciones en los datos.



## Parte Práctica



<center>
<img src="https://pbs.twimg.com/ad_img/1285681293590749189/kDckYy6Z?format=png&name=900x900" width=350 />

Juan Carlos Bodoque, el famoso periodista y empresario, decidió diversificar su portafolio de negocios y crear su propia plataforma de e-commerce. Después de varios años de investigar y analizar el mercado financiero, finalmente logró fundar Bodoque E-Shop con el objetivo de ofrecer a sus clientes una experiencia personalizada y confiable en sus transacciones.

Sin embargo, con la llegada de los aliens al planeta Tierra, aparecen nuevos desafíos para el negocio. Por ello, Bodoque decide invertir en un equipo de expertos en tecnología y comercio interplanetario, para que Bodoque Shop implemente las últimas innovaciones en servicio al cliente para garantizar la satisfacción y fidelización de sus nuevos clientes.

El primer objetivo de Bodoque E-Shop será la hacer un análisis exploratorio para entender mejor el comportamiento de los usuarios en la plataforma. Para ello Bodoque les hace entrega de un extenso dataset en el que se registran las actividades que han realizado sus clientes durante los últimos meses. A continuación se presenta un diccionario de variables que levanto el equipo de consultores interplanetarios de Bodoque:

1. `Transaction ID`: A unique identifier for each transaction.
2. `Customer ID`: A unique identifier for each customer.
3. `Transaction Amount`: The total amount of money exchanged in the transaction in USD.
4. `Transaction Date`: The date and time when the transaction took place.
5. `Payment Method`: The method used to complete the transaction (e.g., credit card, PayPal, etc.).
6. `Product Category`: The category of the product involved in the transaction.
7. `Quantity`: The number of products involved in the transaction.
8. `Customer Age`: The age of the customer making the transaction.
9. `Customer Location`: The geographical location of the customer.
10. `Device Used`: The type of device used to make the transaction (e.g., mobile, desktop).
11. `IP Address`: The IP address of the device used for the transaction.
Shipping Address: The address where the product was shipped.
12. `Billing Address`: The address associated with the payment method.
13. `Is An Alien`: A binary indicator of whether customer is an alien.
14. `Account Age Days`: The age of the customer's account in days at the time of the transaction.
15. `Transaction Hour`: The hour of the day when the transaction occurred.


### Importamos librerias utiles y cargamos los datos😸

In [ ]:
!pip install pyspark
!apt-get install openjdk-8-jdk-headless -qq > /dev/null

import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 317.0/317.0 MB 3.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for pyspark: filename=pyspark-3.5.1-py2.py3-none-any.whl size=317488491 sha256=620c959f2aeab0d04715a8c5c98423b05c9ddeeb85e6798c8969f78facabc38b
  Stored in directory: /root/.cache/pip/wheels/80/1d/60/2c256ed38dddce2fdd93be545214a63e02fbd8d74fb0b7f3a6
Successfully built pyspark


In [ ]:
# Libreria Core del lab.
import pyspark
from pyspark import SparkConf, SparkContext
from pyspark.sql import SparkSession
import pandas as pd
from pyspark.sql.types import StringType, IntegerType, FloatType


#Libreria para plotear
!pip install --upgrade plotly
!pip install missingno
import matplotlib.pyplot as plt
import plotly.express as px

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.7/15.7 MB 54.7 MB/s eta 0:00:00
  Attempting uninstall: plotly
    Found existing installation: plotly 5.15.0
    Uninstalling plotly-5.15.0:
      Successfully uninstalled plotly-5.15.0


Cargue los datos usando **pyspark**

In [ ]:
sparksession = SparkSession.builder.master("local").appName("Laboratorio 4").getOrCreate()

In [ ]:
%%capture
!wget https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/datos_lab_spark.parquet

In [ ]:
df = (
    sparksession.read
    .option("header",'True')
    #.option('delimiter', ',') # Formato del delimiter en el archivo de lectura
    .parquet("datos_lab_spark.parquet")
)

### 2. Limpieza con pyspark [8 puntos]
(1 punto por pregunta)

<center>
<img src="https://miro.medium.com/v2/resize:fit:600/1*A6PpTrehGLxCJWNcUsDTNg.jpeg" width=350 />



Para comenzar con el análisis exploratorio usted decide empezar limpiando la base de datos con **pyspark** dado el alto volumen de datos que genera diariamente Bodoque E-Shop.

**Nota: NO SE PERMITE EL USO DE PANDAS EN ESTA SECCIÓN**



1.   Utilice `.printSchema()` para revisar la estructura de los datos
2.   Muestre las primeras 10 filas del dataset. Hint: utilice `.show()`
3.   Imprima un muestreo aleatorio con el 5% de los datos diponibles. . Hint: utilice `.sample()`
4. Revise los tipos de datos de cada columna con `.dtypes()` y responda la siguiente pregunta: ¿Cuál/es columna/s tiene/n un tipo de dato que no es el adecuado y por qué?
5. Complete el código entregado para cambiar el tipo de datos para la/s columna/s problemáticas.
6. Cuente la cantidad de datos nulos por variable. Recuerde que Spark no posee un método que le permita calcular directamente los nulos.
7. Elimine datos nulos.
8. Elimine datos duplicados.


In [ ]:
# Escriba su respuesta aquí

#Utilice .printSchema() para revisar la estructura de los datos
df.printSchema()

root
 |-- Transaction ID: string (nullable = true)
 |-- Customer ID: string (nullable = true)
 |-- Transaction Amount: double (nullable = true)
 |-- Transaction Date: timestamp_ntz (nullable = true)
 |-- Payment Method: string (nullable = true)
 |-- Product Category: string (nullable = true)
 |-- Quantity: double (nullable = true)
 |-- Customer Age: double (nullable = true)
 |-- Customer Location: string (nullable = true)
 |-- Device Used: string (nullable = true)
 |-- IP Address: string (nullable = true)
 |-- Shipping Address: string (nullable = true)
 |-- Billing Address: string (nullable = true)
 |-- Is An Alien: double (nullable = true)
 |-- Account Age Days: double (nullable = true)
 |-- Transaction Hour: double (nullable = true)



In [ ]:
#Muestre las primeras 10 filas del dataset. Hint: utilice .show()
df.show(10)

+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+-----------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+
|      Transaction ID|         Customer ID|Transaction Amount|   Transaction Date|Payment Method|Product Category|Quantity|Customer Age|Customer Location|Device Used|     IP Address|    Shipping Address|     Billing Address|Is An Alien|Account Age Days|Transaction Hour|
+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+-----------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+
|4b4a5fe1-ec4d-4f9...|d1b87f62-51b2-493...|             58.09|2022-03-24 00:00:00| bank transfer|     electronics|     1.0|        17.0|    Amandaborough|       NULL| 212.195.49.198|Unit 

In [ ]:
#Imprima un muestreo aleatorio con el 5% de los datos diponibles. . Hint: utilice .sample()
df.sample(fraction=0.05, seed=3).show()

+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+------------------+-----------+--------------+--------------------+--------------------+-----------+----------------+----------------+
|      Transaction ID|         Customer ID|Transaction Amount|   Transaction Date|Payment Method|Product Category|Quantity|Customer Age| Customer Location|Device Used|    IP Address|    Shipping Address|     Billing Address|Is An Alien|Account Age Days|Transaction Hour|
+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+------------------+-----------+--------------+--------------------+--------------------+-----------+----------------+----------------+
|                NULL|a748a557-cf3d-481...|              NULL|               NULL|   credit card|   home & garden|     2.0|        13.0|          West Amy|     tablet|123.118.78.180|37175

In [ ]:
#Revise los tipos de datos de cada columna con .dtypes()
df.dtypes

[('Transaction ID', 'string'),
 ('Customer ID', 'string'),
 ('Transaction Amount', 'double'),
 ('Transaction Date', 'timestamp_ntz'),
 ('Payment Method', 'string'),
 ('Product Category', 'string'),
 ('Quantity', 'double'),
 ('Customer Age', 'double'),
 ('Customer Location', 'string'),
 ('Device Used', 'string'),
 ('IP Address', 'string'),
 ('Shipping Address', 'string'),
 ('Billing Address', 'string'),
 ('Is An Alien', 'double'),
 ('Account Age Days', 'double'),
 ('Transaction Hour', 'double')]

**Responda la siguiente pregunta: ¿Cuál/es columna/s tiene/n un tipo de dato que no es el adecuado y por qué?**

- Quantity debiera ser IntegerType -> porque no pueden haber cantidades de objetos comprados decimales

- Is An Alien debiera ser ByteType -> porque es 0 o 1 y no se necesitan decimales ni valores grandes

- Account Age Days debiera ser IntegerType -> La cantidad de días no pueden ser decimales

- Transaction Hour debiera ser IntegerType ->  debido a que la cantidad de horas no son decimales

In [ ]:
#Complete el código entregado para cambiar el tipo de datos para la/s columna/s problemáticas.

#Se extrae función de la Clase-Spark
from pyspark.sql.types import *


def cast_columns(df, cols_types):
    """
    Cambia el tipo de múltiples columnas en un DataFrame de Spark.

    Parámetros:
    - df: DataFrame de Spark.
    - cols_types: Diccionario con nombres de columnas como claves y tipos de datos de Spark como valores.

    Retorna:
    - DataFrame de Spark con tipos de columnas modificados.
    """
    for col_name, new_type in cols_types.items():
        df = df.withColumn(col_name, df[col_name].cast(new_type))
    return df

# Ejemplo de uso
cols_to_cast = {"Quantity": IntegerType(), "Is An Alien": ByteType(), "Account Age Days": IntegerType(), "Customer Age": IntegerType(), "Transaction Hour": IntegerType()}
df = cast_columns(df, cols_to_cast)
df.printSchema()

root
 |-- Transaction ID: string (nullable = true)
 |-- Customer ID: string (nullable = true)
 |-- Transaction Amount: double (nullable = true)
 |-- Transaction Date: timestamp_ntz (nullable = true)
 |-- Payment Method: string (nullable = true)
 |-- Product Category: string (nullable = true)
 |-- Quantity: integer (nullable = true)
 |-- Customer Age: integer (nullable = true)
 |-- Customer Location: string (nullable = true)
 |-- Device Used: string (nullable = true)
 |-- IP Address: string (nullable = true)
 |-- Shipping Address: string (nullable = true)
 |-- Billing Address: string (nullable = true)
 |-- Is An Alien: byte (nullable = true)
 |-- Account Age Days: integer (nullable = true)
 |-- Transaction Hour: integer (nullable = true)



In [ ]:
df.show(10)

+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+-----------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+
|      Transaction ID|         Customer ID|Transaction Amount|   Transaction Date|Payment Method|Product Category|Quantity|Customer Age|Customer Location|Device Used|     IP Address|    Shipping Address|     Billing Address|Is An Alien|Account Age Days|Transaction Hour|
+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+-----------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+
|4b4a5fe1-ec4d-4f9...|d1b87f62-51b2-493...|             58.09|2022-03-24 00:00:00| bank transfer|     electronics|       1|          17|    Amandaborough|       NULL| 212.195.49.198|Unit 

In [ ]:
#Cuente la cantidad de datos nulos por variable.
#Recuerde que Spark no posee un método que le permita calcular directamente los nulos.

#Utilizamos método visto en clases
from pyspark.sql.functions import when, rand, col, sum

exprs = [sum(when(col(c).isNull(), 1).otherwise(0)).alias(c) for c in df.columns]

df.agg(*exprs).show()

+--------------+-----------+------------------+----------------+--------------+----------------+--------+------------+-----------------+-----------+----------+----------------+---------------+-----------+----------------+----------------+
|Transaction ID|Customer ID|Transaction Amount|Transaction Date|Payment Method|Product Category|Quantity|Customer Age|Customer Location|Device Used|IP Address|Shipping Address|Billing Address|Is An Alien|Account Age Days|Transaction Hour|
+--------------+-----------+------------------+----------------+--------------+----------------+--------+------------+-----------------+-----------+----------+----------------+---------------+-----------+----------------+----------------+
|        126886|     126847|            126840|          126831|        126877|          126814|  126966|      126826|           126781|     126911|    126929|          126734|         126804|     126828|          126886|          126953|
+--------------+-----------+----------------

In [ ]:
#Elimine datos nulos.

#Se eliminan cualquier fila que contenga al menos 1 dato nulo
df = df.na.drop(how="any")

In [ ]:
df.show()

+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+--------------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+
|      Transaction ID|         Customer ID|Transaction Amount|   Transaction Date|Payment Method|Product Category|Quantity|Customer Age|   Customer Location|Device Used|     IP Address|    Shipping Address|     Billing Address|Is An Alien|Account Age Days|Transaction Hour|
+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+--------------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+
|bacd9392-73ce-481...|37de64d5-e901-4a5...|            389.96|2023-01-17 00:00:00|    debit card|     electronics|       2|          40|        East Timothy|    desktop|208.106.2

In [ ]:
print((df.count(), len(df.columns)))

(13388, 16)


In [ ]:
#Elimine datos duplicados.
df = df.dropDuplicates(['Transaction ID'])

In [ ]:
print((df.count(), len(df.columns)))

(13237, 16)


In [ ]:
#Se verifica que no queden datos duplicados (extraido de stackoverflow)
if df.count() > df.dropDuplicates(['Transaction ID']).count():
    raise ValueError('Data has duplicates')

### 3. Transformaciones con pyspark [6 puntos]
(1 punto por pregunta)

<center>
<img src="https://live.staticflickr.com/13/91801406_0e71d7f019_b.jpg" width=350 />



**Nota: NO SE PERMITE EL USO DE PANDAS EN ESTA SECCIÓN**

Para continuar con el análisis, los especistas de Bodoque les gustaría tener nuevas variables disponibles. Tras las notas de la reunión usted llega a la conclusión de que tiene que realizar las siguientes tareas (con el dataset preprocesado de la seccion anterior):


1.   Agregar una columna llamada "Transaction bp" con el **monto total** de la transacción en bodoque pesos. Se considera que $x$ dólares equivalen a $log(48+|x^{36}|)$ bodoque pesos.
2.   Crear una columna llamada "Transaction Month" con el mes en que se realiza una transacción.
3.   Crear la variable *Type of purchase* según la catidad de unidades vendidas de acuerdo a las siguientes categorías.
  * Compra minorista: 5 productos o menos.
  * Compra mayorista: 6 produtos o más.
4. Imprima los registros de compras hechas por alienígenas en el comecio mayorista.  Utilice `.filter()`.
5. Cuente la cantidad de compras realizadas por humanos y la cantidad de compras realizadas por alienígenas. Utilice `.groupby()`.
6. Muestre una tabla con la recaudación promedio por transacción para cada método de pago, tanto para humanos como alienígenas. Utilice `pivot()`.

In [ ]:
# Escriba su respuesta aquí
#Agregar una columna llamada "Transaction bp" con el monto total de la transacción
#en bodoque pesos. Se considera que  x  dólares equivalen a  log(48+|x**36|)  bodoque pesos.
from pyspark.sql.functions import abs, log

df=df.withColumn('Transaction bp', log(48+abs(df['Transaction Amount']**36)))

In [ ]:
df.show(5)

+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+-----------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+------------------+
|      Transaction ID|         Customer ID|Transaction Amount|   Transaction Date|Payment Method|Product Category|Quantity|Customer Age|Customer Location|Device Used|     IP Address|    Shipping Address|     Billing Address|Is An Alien|Account Age Days|Transaction Hour|    Transaction bp|
+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+-----------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+------------------+
|0002948c-f473-4fd...|7ae77fa3-3e4c-479...|            131.79|2021-09-22 00:00:00|        PayPal|    toys & games|       2|       

In [ ]:
#Crear una columna llamada "Transaction Month" con el mes en que se realiza una transacción.

from pyspark.sql.functions import month
df=df.withColumn('Transaction Month', month(df['Transaction Date']))

In [ ]:
df.show(5)

+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+-----------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+------------------+-----------------+
|      Transaction ID|         Customer ID|Transaction Amount|   Transaction Date|Payment Method|Product Category|Quantity|Customer Age|Customer Location|Device Used|     IP Address|    Shipping Address|     Billing Address|Is An Alien|Account Age Days|Transaction Hour|    Transaction bp|Transaction Month|
+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+-----------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+------------------+-----------------+
|0002948c-f473-4fd...|7ae77fa3-3e4c-479...|            131.79|2021-09-22 00:

In [ ]:
#Crear la variable Type of purchase según la cantidad de unidades vendidas de acuerdo a las siguientes categorías.
#Compra minorista: 5 productos o menos.
#Compra mayorista: 6 produtos o más.

df=df.withColumn('Type of purchase', when(df['Quantity']<=5, "Compra minorista").otherwise("Compra mayorista"))

In [ ]:
#Se observa que no hay ninguna compra mayorista
df.filter(df['Type of purchase']=='Compra mayorista').show()
df.filter(df['Quantity']>5).show()

+--------------+-----------+------------------+----------------+--------------+----------------+--------+------------+-----------------+-----------+----------+----------------+---------------+-----------+----------------+----------------+--------------+-----------------+----------------+
|Transaction ID|Customer ID|Transaction Amount|Transaction Date|Payment Method|Product Category|Quantity|Customer Age|Customer Location|Device Used|IP Address|Shipping Address|Billing Address|Is An Alien|Account Age Days|Transaction Hour|Transaction bp|Transaction Month|Type of purchase|
+--------------+-----------+------------------+----------------+--------------+----------------+--------+------------+-----------------+-----------+----------+----------------+---------------+-----------+----------------+----------------+--------------+-----------------+----------------+
+--------------+-----------+------------------+----------------+--------------+----------------+--------+------------+---------------

In [ ]:
#Imprima los registros de compras hechas por alienígenas en el comecio mayorista. Utilice .filter().

df.filter(df['Is An Alien']==1).show()

+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+--------------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+------------------+-----------------+----------------+
|      Transaction ID|         Customer ID|Transaction Amount|   Transaction Date|Payment Method|Product Category|Quantity|Customer Age|   Customer Location|Device Used|     IP Address|    Shipping Address|     Billing Address|Is An Alien|Account Age Days|Transaction Hour|    Transaction bp|Transaction Month|Type of purchase|
+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+--------------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+------------------+-----------------+----------------+
|0135e783-77af-4

In [ ]:
#Cuente la cantidad de compras realizadas por humanos y la cantidad de compras realizadas por alienígenas. Utilice .groupby().

df.groupby("Is An Alien").count().show()

+-----------+-----+
|Is An Alien|count|
+-----------+-----+
|          1|  685|
|          0|12552|
+-----------+-----+



In [ ]:
#Muestre una tabla con la recaudación promedio por transacción para
#cada método de pago, tanto para humanos como alienígenas. Utilice pivot()

df.groupby("Payment Method").pivot("Is An Alien").mean("Transaction Amount").show()

+--------------+------------------+-----------------+
|Payment Method|                 0|                1|
+--------------+------------------+-----------------+
|    debit card|215.66604264659748|549.4418644067796|
| bank transfer|  214.751203556685|550.3845930232562|
|        PayPal| 214.8426181936741|469.5177647058822|
|   credit card|206.09774070543347|520.4671686746989|
+--------------+------------------+-----------------+



### 4. EDA [18 puntos]
(1 punto por gráfico y 1 punto por su interpretación)

<center>
<img src="https://i.pinimg.com/originals/41/7e/7b/417e7b9089bcc20c4909df8954c6e742.gif" width=400 />



Esta sección tiene como objetivo evaluar su habilidad para generar reportes y conclusiones a partir de los patrones identificados en los datos proporcionados por Bodoque. Específicamente, se enfoca en **caracterizar las transacciones** y **explorar las diferencias y similitudes en el comportamiento de humanos y aliens**. Utilice el dataset que ya incluye las transformaciones necesarias.

Por favor, asegúrese de que **todas** las visualizaciones que realice cumplan con los siguientes criterios:
- Deben ser relevantes y fáciles de interpretar.
- Cada gráfico debe incluir un título claro, nombres en los ejes y leyendas adecuadas.
- Adjunte una breve descripción interpretativa junto a cada gráfico para explicar los resultados visualizados.

Para llevar a cabo esta tarea, siga los siguientes pasos utilizando la librería de visualización de su elección (matplotlib, seaborn, plotly, etc):

1. **Conversión del DataFrame a formato pandas**: Pase el DataFrame procesado a formato pandas. Evite realizar transformaciones adicionales con pandas.
2. **Visualización de Variables Categóricas**:
   - Genere **tres gráficos de barras** que diferencien entre humanos y aliens. Analice y comente cualquier diferencia o similitud observada entre estos dos grupos.
3. **Visualización de Variables Numéricas**:
   - Elabore **tres distplots** para examinar las distribuciones de variables numéricas, diferenciando entre humanos y aliens. Comente las diferencias o similitudes notables.
4. **Análisis de Patrones en Transacciones**:
   - Cree **tres gráficos avanzados** que ayuden a identificar patrones en las transacciones. Estos gráficos deben incorporar al menos dos dimensiones y diferir de los anteriores. Algunos ejemplos podrían ser un lineplot que muestre la cantidad de transacciones mensuales por canal de venta, o un barplot que exhiba los tres productos más vendidos por canal.

Estos pasos le permitirán no solo visualizar datos complejos de manera efectiva, sino también interpretar estos datos para extraer insights valiosos acerca del comportamiento de los consumidores en el contexto de Bodoque.

In [ ]:
# Escriba su respuesta aquí

#Conversión del DataFrame a formato pandas
df_pd = df.toPandas()

In [ ]:
df_pd

Transaction ID  \
0      0002948c-f473-4fd9-aaf7-6207ff85a889   
1      000ebb9d-d512-40f4-b574-b7e2d010f6ac   
2      001146c5-e77f-4aa9-b1b3-6c0e9d74c376   
3      00180344-94f2-4d4e-98ff-46b73b727b46   
4      001cac84-58c4-4c53-bd60-3f1b1c470e66   
...                                     ...   
13232  fff7aa3c-90eb-40e9-bb36-6947f8317d1b   
13233  fff91a5f-b0f8-4bcb-8b11-8b3d14623045   
13234  fff92c6c-6ed5-4644-a029-853e2ffbf4fd   
13235  fff9d31a-daf7-4ce4-ac08-7e8942578fa9   
13236  fffd85cd-b16e-47aa-a48f-f42d04534372   

                                Customer ID  Transaction Amount  \
0      7ae77fa3-3e4c-4791-a731-f75773e8526a              131.79   
1      9d6ab2e2-aefe-4ed9-84a8-38bce0fa0390              243.12   
2      06d0f47e-080f-4552-a26f-d769c5442153              198.24   
3      f3a873d5-600f-4462-99f5-e744a13f013e              344.53   
4      89d66717-d794-4b7c-95eb-3351051006a5               44.50   
...                                     ...                 ...   
13232  cf8e9612-4c93-4ead-a9e2-08e0034a2c04              255.27   
13233  aab14648-fb5c-4ff5-a44b-c512c4cbed8f              174.03   
13234  53ee5d19-1e24-451b-9fea-a109c1365a4a              385.43   
13235  218bd175-b5d5-43f1-80d7-c46f871191c2              392.15   
13236  f6b6f6d2-77dd-4de9-9145-800f73cea633              375.84   

      Transaction Date Payment Method Product Category  Quantity  \
0           2021-09-22         PayPal     toys & games         2   
1           2023-03-21     debit card      electronics         3   
2           2019-10-29  bank transfer     toys & games         3   
3           2023-08-10         PayPal         clothing         1   
4           2020-09-05    credit card         clothing         3   
...                ...            ...              ...       ...   
13232       2019-06-13    credit card    home & garden         3   
13233       2022-09-13     debit card     toys & games         1   
13234       2018-12-17  bank transfer  health & beauty         2   
13235       2021-06-18    credit card    home & garden         5   
13236       2019-04-02         PayPal  health & beauty         5   

       Customer Age Customer Location Device Used       IP Address  \
0                39         Brownside      tablet    158.98.92.242   
1                37       East Carlos      tablet      73.82.89.46   
2                39         Smithberg     desktop  140.249.125.214   
3                29        Taylorland      mobile  138.123.240.246   
4                34           Penaton      mobile   192.198.69.161   
...             ...               ...         ...              ...   
13232            16     Martinezville      mobile  171.191.201.206   
13233            19        Kirbyhaven      tablet    23.156.135.41   
13234            18       West Robert      tablet   140.150.234.71   
13235            41     West Kimberly      mobile    33.123.64.199   
13236            30     West Colebury      mobile   68.225.241.135   

                                        Shipping Address  \
0      0379 Kerr Valley Apt. 293\nEast Randyburgh, AR...   
1             1752 Reynolds Summit\nEast Linda, DE 34839   
2                                USCGC Cox\nFPO AA 58967   
3                       Unit 3749 Box 6700\nDPO AE 18623   
4      492 John Glens Apt. 005\nNorth Bretttown, IL 1...   
...                                                  ...   
13232  3647 Emily Tunnel Apt. 347\nNorth Angela, FM 1...   
13233          7338 James Course\nCrawfordbury, WY 75809   
13234          31118 Miller Circles\nParkhaven, WA 98043   
13235          0040 Smith Spur\nPort Brianport, DC 08784   
13236                         USCGC Taylor\nFPO AA 29597   

                                         Billing Address  Is An Alien  \
0      0379 Kerr Valley Apt. 293\nEast Randyburgh, AR...            0   
1             1752 Reynolds Summit\nEast Linda, DE 34839            0   
2                                USCGC Cox\nFPO AA 58967     

In [ ]:
df_pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13237 entries, 0 to 13236
Data columns (total 19 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Transaction ID      13237 non-null  object        
 1   Customer ID         13237 non-null  object        
 2   Transaction Amount  13237 non-null  float64       
 3   Transaction Date    13237 non-null  datetime64[ns]
 4   Payment Method      13237 non-null  object        
 5   Product Category    13237 non-null  object        
 6   Quantity            13237 non-null  int32         
 7   Customer Age        13237 non-null  int32         
 8   Customer Location   13237 non-null  object        
 9   Device Used         13237 non-null  object        
 10  IP Address          13237 non-null  object        
 11  Shipping Address    13237 non-null  object        
 12  Billing Address     13237 non-null  object        
 13  Is An Alien         13237 non-null  int8      

In [ ]:
#Visualización de Variables Categóricas:
#Genere tres gráficos de barras que diferencien entre humanos y aliens.
#Analice y comente cualquier diferencia o similitud observada entre estos dos grupos

In [ ]:
df_pd = pd.concat(
    [
        df_pd.select_dtypes([], ["object"]),
        df_pd.select_dtypes(["object"]).apply(pd.Series.astype, dtype="category"),
    ],
    axis=1,)

In [ ]:
numeric = df_pd.select_dtypes(["float64", "int32", "int8"]).columns
categorical = df_pd.select_dtypes(["category"]).columns
dates = df_pd.select_dtypes(["datetime64[ns]"]).columns

In [ ]:
numeric

Index(['Transaction Amount', 'Quantity', 'Customer Age', 'Is An Alien',
       'Account Age Days', 'Transaction Hour', 'Transaction bp',
       'Transaction Month'],
      dtype='object')

In [ ]:
df_pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13237 entries, 0 to 13236
Data columns (total 19 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Transaction Amount  13237 non-null  float64       
 1   Transaction Date    13237 non-null  datetime64[ns]
 2   Quantity            13237 non-null  int32         
 3   Customer Age        13237 non-null  int32         
 4   Is An Alien         13237 non-null  int8          
 5   Account Age Days    13237 non-null  int32         
 6   Transaction Hour    13237 non-null  int32         
 7   Transaction bp      13237 non-null  float64       
 8   Transaction Month   13237 non-null  int32         
 9   Transaction ID      13237 non-null  category      
 10  Customer ID         13237 non-null  category      
 11  Payment Method      13237 non-null  category      
 12  Product Category    13237 non-null  category      
 13  Customer Location   13237 non-null  category  

In [ ]:
fig = px.histogram(df_pd, x="Payment Method", barmode="group", color="Is An Alien",
                   title='Distribución de métodos de pago utilizados por humanos y aliens',
                   labels={'Payment Method': 'Método de Pago', 'count': 'Cantidad'})
fig.show()

- En el histograma es posible visualizar que todos los métodos de pagos presentan una distribución y cantidad de compras similares tanto en compras realizadas por aliens como por humanos.

- Si bien la diferencia de cantidades no es amplia, particularmente existe una mayor cantidad de humanos que compra con tarjeta de debito, lo que también se da en los aliens. Por otro lado, la menor cantidad de compras de humanos es por Paypal, mientras que se observa que menor cantidad de aliens compran con tarjeta de crédito.

In [ ]:
fig = px.histogram(df_pd, x="Product Category", barmode="group", color="Is An Alien",
                   title='Distribución de categorías compradas por humanos y aliens',
                   labels={'Product Category': 'Categoría del producto'})
fig.show()

- En el caso de las categorías por producto comprados, se visualiza un comportamiento semejante al del gráfico de arriba en el que la diferencia por cada categoría no es significativa.

- Asimismo, se distingue que tanto los humanos como los aliens presentan una mayor cantidad de compras en productos eletronicos. Por otro lado, los humanos compran menor cantidad de juegos y juguetes, mientras que los aliens tienen menos compras en ropas.

In [ ]:
fig = px.histogram(df_pd, x="Device Used", barmode="group", color="Is An Alien",
                   title='Distribución de dispositivos utilizados por humanos y aliens',
                   labels={'Device Used': 'Dispositivo utilizado'})
fig.show()

- Para este último gráfico categórico de comparación entre los aliens y los humanos, se puede distinguir un comportamiento similar al obtenido en los gráficos anteriores, donde las distribuciones no presentan mayores diferencias.

- En específico, los aliens presentan mayor cantidad de compras realizadas a través del computador, mientras que los humanos en compras realizadas a través de la tablet. Mientras que ocurre el caso inverso cuando de la menor cantidad de dispositivos utilziados se trata, es decir, menos cantidad de aliens compraron con tablet y menor cantidad de humanos compraron con el computador.

In [ ]:
#Visualización de Variables Numéricas:
#Elabore tres distplots para examinar las distribuciones de variables numéricas,
#diferenciando entre humanos y aliens. Comente las diferencias o similitudes notables.

In [ ]:
px.violin(df_pd, x="Transaction Amount", color="Is An Alien",
          title='Distribución de los montos de las transferencias por humanos y aliens',
          labels={'Transaction Amount': 'Monto de Transferencia'})

- En este gráfico se puede visualizar la distribución comparativa de las transferencias por humanos y por aliens.

 - El gráfico correspondiente a los aliens muestra una variabilidad significativamente mayor, con una mediana aproximada de 136 y transferencias que alcanzan valores negativos hasta -162. Por otro lado, la distribución de las transferencias de los humanos es más acotada, con montos que tienden a agruparse alrededor de los 50. Es notable que los aliens, en comparación con los humanos, tienden a realizar transferencias de montos superiores, llegando incluso a superar los 4600.

In [ ]:
px.histogram(df_pd, x="Transaction bp",  marginal='box', barmode="overlay", color="Is An Alien",
            title='Distribución de los montos de las transferencias en bodoque pesos por humanos y aliens',
          labels={'Transaction bs': 'Transferencia en Bodoque Pesos'})

- Se puede observar que al pasar la transferencia a un monto en bodoque pesos la diferencia observada en la gráfica anterior se hace menos visible, en otras palabras la escala se empequeñece haciendo menos evidente si la compra la podría haber realizado un alien o un humano.

- Aún así, es posible observar diferencias en los valores medios y en los cuartiles, lo que podría ayudar a diferenciar ambas muestras. Se puede distinguir que las transferencias de los aliens son en gran parte valores de mayor magnitud, presentando una mediana cerca a 198 bodoque pesos, en contraste al de los humanos de 180 bodoque pesos.

In [ ]:
px.histogram(df_pd, x="Customer Age",  marginal='box', barmode="overlay", color="Is An Alien",
              title='Distribución de las edades de los clientes por humanos y aliens',
              labels={'Customer Age': 'Edad clientes'})

- En cuanto a la edad de los clientes, tanto para humanos como para aliens tienen una distribución ampliamente similar, siendo la mediana y el extremo del tercer cuartil igual para ambos, de 34 y 41 años respectivamente.

- En este caso, los valores difieren en mayor medida en los extremos, teniendo los clientes humanos menor y mayor edad en casos aislados que los aliens. Sin embargo, no es una diferencia realmente significativa.

In [ ]:
#Análisis de Patrones en Transacciones:
#Cree tres gráficos avanzados que ayuden a identificar patrones en las transacciones.
#Estos gráficos deben incorporar al menos dos dimensiones y diferir de los anteriores.
#Algunos ejemplos podrían ser un lineplot que muestre la cantidad de transacciones
#mensuales por canal de venta, o un barplot que exhiba los tres productos más vendidos por canal.

In [ ]:
df_per_device = df_pd.groupby(['Device Used', 'Product Category'])['Transaction ID'].count().reset_index(name='Count')

fig = px.histogram(df_per_device, x="Device Used", y='Count',  barmode="group", color="Product Category",
                   title='Cantidad de transacciones por canal de venta según dispositivo utilizado',
                   labels={'Device Used': 'Dispositivo utilizado', 'Count': 'Cantidad de transacciones'})
fig.show()

* Del gráfico anterior se puede observar que el mayor
 número de transacciones es realizado por los telefonos de los usuarios, donde destaca la categoría **toys & games** por sobre las demás. Caso contrario ocurre en las transacciones realizadas por tablet o desktop, donde la categoría de **toys & games** es la que menos genera transacciones.

 En cuanto a las compras realizadas por desktop, la categoría con mayor número de transacciones es **electronics** junto con **home & garden**. Mientras que por tables, el número de transacciones por categoría se mantiene parejo para cada una, destacando **home & garden**.

In [ ]:
df_per_month = df_pd.groupby(['Transaction Month', 'Payment Method'])['Transaction ID'].count().reset_index(name='Count')

fig = px.line(df_per_month, x='Transaction Month', y='Count', color='Payment Method',
              title='Cantidad de transacciones mensuales por canal de venta',
              labels={'Transaction Month': 'Mes', 'Count': 'Cantidad de transacciones'})

fig.show()

* En cuanto a la cantidad de transacciones mensuales
 por canal de venta, se puede ver que en enero se alcanza la mayor cantidad de transaciones realizadas por un medio de pago, correspondiente a **Tarjeta de Debito**.

 En febrero se puede ver que hay una baja de transacciones para cada uno de los métodos a excepción de **PayPal** que aumenta levemente.
 Luego, en marzo se puede ver un aumento significativo del número de transacciones por método, alcanzando el segundo peak de transacciones realizadas con **Tarjeta de Credito**.

 Durante lo que queda del primer semestre se puede ver una disminución del número de transacciones principalmente de **Paypal** y **Tarjeta de Credito**.

 Luego en el segundo semestre las disminuciones se observan para **Tarjeta de Debito** y **Paypal**, mientras que **Transferencia bancaría** permanece fluctuante.

In [ ]:
df_per_age_count = df_pd.groupby(['Customer Age', 'Product Category'])['Transaction ID'].count().reset_index(name='Count')

fig = px.line(df_per_age_count, x='Customer Age', y='Count', color='Product Category',
              title='Cantidad de transacciones mensuales por edad de clientes',
              labels={'Customer Age': 'Edad', 'Count': 'Cantidad de transacciones'})

fig.show()

* En el gráfico se puede observar que la curva de distribución de las transacciones por edad de los clientes, es similar para cada una de las diferentes categorías de productos, como lo son Clothing, Electronics, Health & beauty, etc.

 Además, se puede observar claramente que los usuarios entre 30 y 40 años son los que han realizado el mayor número de transacciones dentro de la data histórica, para cada una de las categorías. En específico, la categoría **Electronics** alcanza el mayor peak de transacciones a los 33 años, luego sigue **Toys & games** a los 35 años, **Clothing** y **Health & beauty** a los 36 años, y **Home & garden** a los 31 años.


In [ ]:
fig = px.histogram(df_pd, x="Transaction Month", barmode="group", color="Product Category",
                   title='Cantidad de transacciones mensuales por categoría del producto',
                   labels={'Transaction Month': 'Mes', 'Count': 'Cantidad de transacciones'})
fig.show()

* Se puede observar que en los meses de enero, mayo,
 julio, noviembre y diciembre la categoría con mayor número de transacciones es **electronics**, siendo la categoría con mayor popularidad a lo largo del año.

 Luego sigue la categoría de **health &beauty** predominante en los meses de abril, agosto y octubre.Junto con **toys & games** en los meses de febrero y marzo.

 En cuanto a las categorías **home & garden** y **clothing** alcanzan el mayor número de transacciones unicamente un mes en el año, que corresponden a junio y septiembre respectivamente.


### 5. Particiones y consultas en SQL [6 puntos]
(2 puntos por pregunta)

<center>
<img src="https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/misc_images/1696330143457.gif" width=400 />

El equipo de Bodoque e-shop ha solicitado que los datos estén disponibles en una tabla SQL consultable. Además, están interesados en aprovechar las funciones de ventana en SQL para análisis avanzados. Las funciones de ventana permiten realizar cálculos sobre un conjunto de filas que están relacionadas con la fila actual. Por ejemplo, UNBOUNDED PRECEDING se usa para indicar que el rango de la función de ventana comienza desde la primera fila de la partición o del conjunto de resultados, lo cual es útil para calcular sumas acumulativas hasta la fila actual. Las variaciones comunes de este uso incluyen:

- `UNBOUNDED PRECEDING` to `CURRENT ROW`: Calcula desde el inicio de la partición hasta la fila actual.
- `UNBOUNDED PRECEDING` to `UNBOUNDED FOLLOWING`: Cubre todas las filas dentro de la partición.
- `VALUE PRECEDING` to `VALUE FOLLOWING`: Establece un rango específico basado en valores antes y después de la fila actual.

<center>
<img src="https://learnsql.com/blog/sql-window-functions-rows-clause/1.png" width=500 />

Ejemplo de uso en SQL:

```sql
STAT(COL1_NAME) OVER (PARTITION BY COL2_NAME ORDER BY COL3_NAME ROWS BETWEEN X PRECEDING AND CURRENT ROW)
```


Responda y realice los siguientes puntos:

1. **Creación de Tabla con PySpark**:
   - Desarrolle un script utilizando PySpark para crear una tabla a partir de un DataFrame previamente transformado. Seleccione y utilice una variable específica para la partición de la tabla. Justifique su elección de esta variable considerando factores como el tamaño del DataFrame, la distribución de los datos y el impacto potencial en el rendimiento de futuras consultas.

2. **Consulta SQL para Principales Clientes**:
   - Ejecute una consulta SQL para identificar los 10 clientes que más productos han comprado. La consulta debe retornar el ID del cliente junto con el total de productos comprados, ordenados en forma descendente.

3. **Implementación de Función de Ventana en SQL y Equivalente en Spark**:
   - Implemente una función de ventana en SQL para calcular la compra más alta realizada por cada usuario en los últimos tres meses. Además, describa cómo se podría realizar una función equivalente en Spark, considerando las capacidades específicas de PySpark para manejar este tipo de consultas.



In [ ]:
#Se cambia el nombre de la columna necesitada porque tira error de sintaxis
df = df.withColumnRenamed('Customer ID','Customer_ID')

In [ ]:
df.write.partitionBy("Product Category", "Payment Method").saveAsTable(name='Lab4')

In [ ]:
sparksession.sql("""
    SELECT Customer_ID, SUM(Quantity) AS Total_Products_Bought
    FROM Lab4 GROUP BY Customer_ID
    ORDER BY Total_Products_Bought DESC
    LIMIT 10""").show()

+--------------------+---------------------+
|         Customer_ID|Total_Products_Bought|
+--------------------+---------------------+
|ddb1eedb-78a7-403...|                    5|
|6a6e91d8-d630-466...|                    5|
|ddfbe502-d51c-4ac...|                    5|
|ebf10b43-2999-4fa...|                    5|
|e91ebafd-1925-400...|                    5|
|a76f00f5-8493-41b...|                    5|
|2f3d5b31-bd85-4d7...|                    5|
|6def79d9-fb71-415...|                    5|
|aa8c72f5-fbcc-4bf...|                    5|
|62b1ce0c-5495-43e...|                    5|
+--------------------+---------------------+



In [ ]:
#Implementación de Función de Ventana en SQL y Equivalente en Spark

#Consulta en SQL

SELECT
    Customer_ID,
    MAX(Transaction Amount) AS Max_Amount
FROM
    part_table
WHERE
    Transaction Date >= DATE_SUB(CURRENT_DATE, 90)
GROUP BY
    Customer_ID;

In [ ]:
df2 = df.withColumnRenamed('Transaction Amount','Transaction_Amount')
df2 = df2.withColumnRenamed('Transaction Date','Transaction_Date')

In [ ]:
df2.write.partitionBy("Product Category", "Payment Method").saveAsTable(name='func_vent')

In [ ]:
from pyspark.sql.functions import date_sub, current_date

#Como usarla con PySpark

sparksession.sql("""
SELECT
    Customer_ID,
    MAX(Transaction_Amount) AS Max_Amount
FROM
    func_vent
WHERE
    Transaction_Date >= date_sub(current_date(), 90)
GROUP BY
    Customer_ID;""").show()

+--------------------+----------+
|         Customer_ID|Max_Amount|
+--------------------+----------+
|2f124388-809c-4a1...|    194.72|
|13df5d31-0e77-4e8...|      74.2|
|5d8859e2-c2b5-4b7...|    293.49|
|33dea35f-1ea2-4e5...|     32.02|
|09e610e6-394a-4b7...|    165.94|
|d47fd061-c3e3-428...|      63.6|
|3f1dd40e-7a14-495...|     23.56|
|3f19a9b6-4669-41a...|      39.2|
|a52976f6-b05e-4ff...|    110.25|
|a1f61636-d50b-434...|     311.8|
|60d6c87d-4a7f-49f...|    204.12|
|66402894-b988-4fc...|     77.06|
|b575355f-4ebc-40a...|      95.5|
|4d9b6e88-d044-470...|     661.2|
|72d186f6-82dd-4d7...|     80.19|
|07e9ecc2-b892-432...|    151.07|
|6e58b5c7-70c6-42e...|    299.35|
|298de15c-61ed-42c...|     48.33|
|d25282ce-5ef5-4ea...|     22.89|
|599ee2e2-482e-47b...|    105.33|
+--------------------+----------+
only showing top 20 rows



### 6. UDF [10 puntos]
(2 por pregunta)

<center>
<img src="https://64.media.tumblr.com/ba8c705edd2bed0a28d9458811155d69/tumblr_pap19zg4ae1w3zg6go1_400.gifv" width=400 />



Un experto en predicciones y programación le ha proporcionado un objeto serializado (`pickle`) diseñado para calcular las probabilidades de que un cliente cometa o no un fraude. Este experto sugiere que, para maximizar las capacidades de procesamiento distribuido de Spark, debería implementar `Scalar User Defined Functions` (udf). Esto le permitirá aplicar el objeto serializado en un entorno distribuido a lo largo de toda la población de datos. Un aspecto clave de la función desarrollada por el experto es que se enfoca exclusivamente en las siguientes columnas para realizar las predicciones: `['Transaction Amount', 'Quantity', 'Customer Age', 'Transaction Hour']`.

Aparte, el experto le proporciona las siguientes instrucciones para usar las UDF en Spark:

```python
from pyspark.sql.functions import udf
from pyspark.sql.types import FloatType

def custom_function(col):
    pass

udf_function = udf(custom_function, FloatType())
```

Basándose en la estructura proporcionada, debe desarrollar una función que ejecute un código específico. Tenga en cuenta que esta función solo puede recibir columnas de Spark y debe retornar el valor deseado. Posteriormente, deberá utilizar esta función UDF indicando la función personalizada y el formato de salida.

Siga los siguientes pasos para implementar la solución y responda las preguntas:

1. **Cargar el objeto serializado**: Revise el tipo de objeto y deduzca su función.
2. **Explorar el objeto**: Utilice las funciones `dir` y `help` para identificar qué método del objeto predice la probabilidad.
3. **Crear una función personalizada**: Elabore una función que prediga la probabilidad de fraude utilizando el último valor de la lista generada por el objeto serializado. Puede modificar el nombre de la función para reflejar su propósito.
4. **Definir la función UDF**: Establezca la función UDF con la función personalizada que ha creado.
5. **Generar una nueva columna**: Añada una nueva columna `prediction` a su DataFrame en Spark utilizando la función UDF y muestre un ejemplo de cómo se aplica. ¿Qué beneficios podría generar utilizar udf?

In [ ]:
%%capture
!wget https://gitlab.com/imezadelajara/datos_clase_7_mds7202/-/raw/main/object.pkl?inline=false

In [ ]:
import pickle

#Cargar el objeto serializado: Revise el tipo de objeto y deduzca su función.
pickle_path = "/content/object.pkl?inline=false"
with open(pickle_path, "rb") as f:
    loaded_object = pickle.load(f)

print("Tipo de objeto cargado:", type(loaded_object))

Tipo de objeto cargado: <class 'sklearn.linear_model._logistic.LogisticRegression'>


Es un modelo que realiza Regresión Logística

In [ ]:
#Explorar el objeto: Utilice las funciones dir y help para identificar
#qué método del objeto predice la probabilidad.

help(loaded_object)

Help on LogisticRegression in module sklearn.linear_model._logistic object:

class LogisticRegression(sklearn.linear_model._base.LinearClassifierMixin, sklearn.linear_model._base.SparseCoefMixin, sklearn.base.BaseEstimator)
 |  LogisticRegression(penalty='l2', *, dual=False, tol=0.0001, C=1.0, fit_intercept=True, intercept_scaling=1, class_weight=None, random_state=None, solver='lbfgs', max_iter=100, multi_class='auto', verbose=0, warm_start=False, n_jobs=None, l1_ratio=None)
 |  
 |  Logistic Regression (aka logit, MaxEnt) classifier.
 |  
 |  In the multiclass case, the training algorithm uses the one-vs-rest (OvR)
 |  scheme if the 'multi_class' option is set to 'ovr', and uses the
 |  cross-entropy loss if the 'multi_class' option is set to 'multinomial'.
 |  (Currently the 'multinomial' option is supported only by the 'lbfgs',
 |  'sag', 'saga' and 'newton-cg' solvers.)
 |  
 |  This class implements regularized logistic regression using the
 |  'liblinear' library, 'newton-cg', '

In [ ]:
dir(loaded_object)

['C',
 '__annotations__',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_check_feature_names',
 '_check_n_features',
 '_estimator_type',
 '_get_param_names',
 '_get_tags',
 '_more_tags',
 '_parameter_constraints',
 '_predict_proba_lr',
 '_repr_html_',
 '_repr_html_inner',
 '_repr_mimebundle_',
 '_validate_data',
 '_validate_params',
 'class_weight',
 'classes_',
 'coef_',
 'decision_function',
 'densify',
 'dual',
 'feature_names_in_',
 'fit',
 'fit_intercept',
 'get_params',
 'intercept_',
 'intercept_scaling',
 'l1_ratio',
 'max_iter',
 'multi_class',
 'n_features_in_',
 'n_iter_',
 'n_jobs',
 'penalty',
 'predict',
 'predict_log_

¿qué método del objeto predice la probabilidad?
Método predict_proba

In [ ]:
# Código Aquí

from pyspark.sql.functions import udf
from pyspark.sql.types import FloatType

#Crear una función personalizada: Elabore una función que prediga la probabilidad
#de fraude utilizando el último valor de la lista generada por el objeto serializado.
#Puede modificar el nombre de la función para reflejar su propósito.

def predict_fraud(vals):
    vals = [vals]  # Convertir a array 2D
    predict_probs = loaded_object.predict_proba(vals)
    last = predict_probs[0][-1]
    #print(last)
    return float(last)

#Definir la función UDF: Establezca la función UDF con la función personalizada que ha creado.
udf_function = udf(predict_fraud, FloatType())

In [ ]:
#Generar una nueva columna: Añada una nueva columna prediction a su DataFrame en
#Spark utilizando la función UDF y muestre un ejemplo de cómo se aplica.
from pyspark.sql.functions import array

columns_array = array('Transaction Amount', 'Quantity', 'Customer Age', 'Transaction Hour')


df_predictfraud = df.withColumn('Fraud Predictions', udf_function(columns_array))

In [ ]:
df_predictfraud.show(5)

+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+-----------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+------------------+-----------------+----------------+-----------------+
|      Transaction ID|         Customer_ID|Transaction Amount|   Transaction Date|Payment Method|Product Category|Quantity|Customer Age|Customer Location|Device Used|     IP Address|    Shipping Address|     Billing Address|Is An Alien|Account Age Days|Transaction Hour|    Transaction bp|Transaction Month|Type of purchase|Fraud Predictions|
+--------------------+--------------------+------------------+-------------------+--------------+----------------+--------+------------+-----------------+-----------+---------------+--------------------+--------------------+-----------+----------------+----------------+------------------+-----------------+-------

In [ ]:
sparksession.stop()

**¿Qué beneficios podría generar utilizar udf?**

Se puede mejorar la rapidez de procesamiento al aplicar una función dada en toda una columna de una sola vez. Además, entrega flexibilidad al aplicar modelos dentro de los datos de un dataframe, como lo son las columnas que recibe en este caso.
